joomla/joomla-cms
# Bundled FOF 2.4.3 is out of date and incompatible with PHP 7

Pull Request for Issue #11672 .

Pinging @wilsonge and @mbabker – This PR addresses PHP 7.0 and 7.1 compatibility issues. See the bold stuff under Bugfixes
### Summary of Changes

Updated FOF to 2.5.5. The following improvements and bug fixes have been implemented since version 2.4.3 (what is currently included in Joomla! 3.6.2):
#### Improvements
- Now compatible with Joomla! 1.6, 1.7, 2.5, 3.0, 3.1, 3.2, 3.3, 3.4 and 3.5 (best used with Joomla! 3.3 or later)
- F0FUtilsIniFile to parse INI files even on broken hosts
- F0FUtilsConfigHelper to get component configuration parameters without going through JComponentHelper
- F0FUtilsUpdate improvements
- Clear the OPcache when installing / updating an extension as long as it uses F0FUtilsInstallscript
- F0FUtilsIp: let the workarounds be disabled
- gh-584 removeHeader method added in Form, allowing you to remove a header from an XML form.
- utf8mb4 support in schema installer under Joomla! 3.5+
#### Bugfixes
- [HIGH] **mcrypt is deprecated in PHP 7.1**. This will cause Two Factor Authentication to fail under PHP 7.1 unless you apply this PR.
- [HIGH] **gh-593 Cannot use multiple behaviours under PHP 7**. This was breaking many FOF extensions using the old version in Joomla! 3.5 and 3.6 under PHP 7.
- [HIGH] If you declared onBefore/onAfter Controller methods public FOF would erroneously treat them as tasks
- [HIGH] Crazy code for JDatabaseInterface in Joomla! causes fatal errors when using the Joomla! db driver in FOF in CLI
- [HIGH] The caching ID was always wrong since JCache::makeId() is never called any more in Joomla!.
- [LOW] F0FUtilsUpdate had a wrong version_compare which didn't allow the automatic update to proceed
- [HIGH] F0FUtilsUpdateJoomla could not detect Joomla! updates due to changes in Joomla! XML update sources
### Testing Instructions

After applying the PR make sure that the two core features using FOF still work.
1. Two Factor Authentication feature works. Espeially for @brainteeman : If you don't remember how TFA works please RTFM that I kindly provided back in 2012 https://docs.joomla.org/J3.x:Two_Factor_Authentication
2. The Post-installation Messages component. Install a BRAND NEW copy of Joomla!. Log in to the Administrator. You will see some post-installation messages. Make sure you can still dismiss them. Especially for @brianteeman : Since that core component may still have bugs and you have a propensity to bumping into bugs please make sure that any bugs you spot are NOT reproducible in plain old Joomla! 3.6.2. If they are please file a different GitHub issue and at-mention me.
### Documentation Changes Required

None.
